### PR TITLE
RFC for testing : [Xen 4.9] Upgrade Xen hypervisor and toolstack to 4.9.0

### PR DESCRIPTION
**Pull request by dozylynx**
OXT-1203 : This PR is submitted as a RFC, provided to enable testing and evaluation of using Xen 4.9 with OpenXT.

There are significant changes to the OpenXT patches in the Xen queue and in the large upstream body of code that this PR upgrades.

Signed-off-by: Christopher Clark <christopher.clark6@baesystems.com>

### CODE REVIEW

**Comment by jean-edouard** [comment]
Built here: http://openxt-builder.ainfosec.com:8010/builders/openxt/builds/1078

**Comment by jean-edouard** [comment]
This is the list of PRs that go with this one:
https://github.com/OpenXT/manager/pull/97
https://github.com/OpenXT/openxt/pull/265
https://github.com/OpenXT/surfman/pull/21
https://github.com/OpenXT/toolstack/pull/29
https://github.com/OpenXT/xsm-policy/pull/15

**Comment by jean-edouard** [comment]
The other PRs seem pretty minor, and all look good to me.

**Comment by dozylynx** [comment]
> When testing, I found that guest networking and atapi passthrough don't seem to work.
@dozylynx, do you know about those issues, are there other known issues and are you currently working on any of them? I'm just trying to make sure we don't duplicate efforts.

Thanks for looking at this -- and no, I hadn't tested those. I got the system to the point that it boots, has XSM and SELinux enforcing, can install Windows 7 64bit without tools, have the display and host networking working correctly, and fixed what I'd found on the way to get it to that point. From there, it works well enough for my immediate needs to be able to do development work with, so I haven't tested it further than that. This is pretty much a long way of saying: I don't have a list! so please go ahead.

File: recipes-core/images/xenclient-installer-image/isolinux.cfg [hunk 1]
```diff
@@ -5,10 +5,10 @@ PROMPT 1
 TIMEOUT 20
 LABEL xenclient
   kernel mboot.c32
-  append tboot.gz min_ram=0x2000000 loglvl=all serial=115200,8n1,0x3f8 logging=serial,memory --- xen.gz console=com1 dom0_max_vcpus=1 com1=115200,8n1,pci dom0_mem=max:8G ucode=-1 --- vmlinuz quiet root=/dev/ram rw start_install=new eject_cdrom=1 answerfile=/install/answers/default.ans console=hvc0 console=/dev/tty2 selinux=0 --- rootfs.gz --- gm45.acm --- q35.acm --- q45q43.acm --- duali.acm --- quadi.acm --- ivb_snb.acm --- xeon56.acm --- xeone7.acm --- hsw.acm --- bdw.acm --- skl.acm --- kbl.acm --- microcode_intel.bin
+  append tboot.gz min_ram=0x2000000 loglvl=all serial=115200,8n1,0x3f8 logging=serial,memory --- xen.gz flask=disabled console=com1 dom0_max_vcpus=1 com1=115200,8n1,pci dom0_mem=max:8G ucode=-1 --- vmlinuz quiet root=/dev/ram rw start_install=new eject_cdrom=1 answerfile=/install/answers/default.ans console=hvc0 console=/dev/tty2 selinux=0 --- rootfs.gz --- gm45.acm --- q35.acm --- q45q43.acm --- duali.acm --- quadi.acm --- ivb_snb.acm --- xeon56.acm --- xeone7.acm --- hsw.acm --- bdw.acm --- skl.acm --- kbl.acm --- microcode_intel.bin
```
Comment (dozylynx) [inline recipes-core/images/xenclient-installer-image/isolinux.cfg]: To be clear, this change is just for the host installer, not for the resulting system. An upstream change is that XSM is enabled by default in this version of Xen, whereas it wasn't in 4.6. On the machines that I tested on it just aborted boot and this was the fastest way to get past that. As far as I'm aware, we don't have a suitable XSM policy for the installer to load, so it didn't seem like a particularly negative change.

File: recipes-openxt/xenclient/xenmgr/files/xenstore-init-extra [hunk 2]
```diff
@@ -1,5 +1,8 @@
 #!/bin/sh
 
+# xl toolstack domain id
+xenstore-write 'domid' 0
```
Comment (dozylynx) [inline recipes-openxt/xenclient/xenmgr/files/xenstore-init-extra]: I wasn't familiar with that tool, but yes, it looks like it is intended to do exactly that.

File: recipes-extended/xen/files/libxl-xenmgr-support.patch [hunk 3]
```diff
@@ -32,147 +34,125 @@ INTERNAL DEPENDENCIES
...
-@@ -3525,6 +3538,8 @@ void libxl__device_nic_add(libxl__egc *e
-     aodev->action = LIBXL__DEVICE_ACTION_ADD;
-     libxl__wait_device_connection(egc, aodev);
+--- xen-4.9.0.orig/tools/libxl/libxl.h
++++ xen-4.9.0/tools/libxl/libxl.h
+@@ -1405,6 +1405,7 @@ int libxl_domain_remus_start(libxl_ctx *
+                              const libxl_asyncop_how *ao_how)
+                              LIBXL_EXTERNAL_CALLERS_ONLY;
  
-+    libxl__xs_write(gc, XBT_NULL, GCSPRINTF("%s/hotplug-status", libxl__device_backend_path(gc, device)), "connected");
```
Comment (jean-edouard) [inline recipes-extended/xen/files/libxl-xenmgr-support.patch]: Seems like this is lost for NICs, which is what I think breaks pv networking.

Comment (jean-edouard) [inline recipes-extended/xen/files/libxl-xenmgr-support.patch]: I have a fix for this, but wasn't allowed to push to your branch.
Could you give me access? If you'd rather not, I don't mind merging this and writing a new PR for my fix.

Comment (dozylynx) [inline recipes-extended/xen/files/libxl-xenmgr-support.patch]: fix: cool, thanks.
I do have a preference for a separate PR for the edit because I use the single URL for each PR to cover IP ownership reporting and it would create more work to clarify separate contributions. Hope that's ok.

Comment (jean-edouard) [inline recipes-extended/xen/files/libxl-xenmgr-support.patch]: Yup, no problem, this is the fix I'll PR once this is merged:
https://github.com/jean-edouard/xenclient-oe/commit/3b2fdbc72d8b4dfbe8072bf31d85340f327c9882

**Comment by jean-edouard** [comment]
TL;DR: merging tomorrow by end of day.
Despite the name of this PR, this upgrade is actually pretty solid!
As noted above, I found a couple of small issues, but I have fixes ready for them.
Since Github doesn't support stacking PRs, I'll have to wait until after everything is merged to submit my fixes.
I'm sure more in-depth testing will uncover some issues, but master doesn't have to be super stable, and we definitely want/need to upgrade Xen for the next major release.
With that in mind, I suggest we merge this PR and the related ones (see above for the list) tomorrow towards the end of the business day, and I will do so if nobody objects by then.
Thanks @dozylynx!

**Comment by richpersaud** [comment]
Sounds good.  For any issues found, please file tickets under the Xen uprev epic, https://openxt.atlassian.net/browse/OXT-1203
